pkg/gohai/utils/common.go

```diff
@@ -0,0 +1,183 @@
+	if !secondRetType.Implements(reflErrorInterface) {
+		return reflect.Method{}, false
+	}
+
+	return valueMethod, true
+}
+
+// AsJSON takes an Info struct and returns a marshal-able object representing the fields of the struct,
+// the lists of errors for fields for which the collection failed, and an error if it failed.
+//
+// If useDefault is true, fields which failed to be collected will be included in the marshal-able object
+// as their default value, otherwise they are ignored.
+//
+// Fields which are not exported, don't have a json tag or are not of type Value[T] for a T which can
+// be rendered cause the function to return an error.
+//
+// The string list contain errors of fields which failed to be collected. Fields which are not
+// collected on the platform are ignored.
+//
+// If the error is non-nil, the first two parameters are unspecified.
+func AsJSON[T any](info *T, useDefault bool) (interface{}, []string, error) {
+	reflVal := reflect.ValueOf(info).Elem()
+	reflType := reflect.TypeOf(info).Elem()
+
+	// info has to be a struct
+	if reflVal.Kind() != reflect.Struct {
+		return nil, nil, ErrArgNotStruct
+	}
+
+	values := make(map[string]interface{})
+	warns := []string{}
+
+	for i := 0; i < reflVal.NumField(); i++ {
+		fieldTy := reflType.Field(i)
+		fieldName := fieldTy.Name
+
+		// check if field is exported
+		if !fieldTy.IsExported() {
+			return nil, nil, fmt.Errorf("%s: %w", fieldName, ErrNotExported)
+		}
+
+		// check if field has a json tag
+		jsonName, ok := fieldTy.Tag.Lookup("json")
+		if !ok {
+			return nil, nil, fmt.Errorf("%s: %w", fieldName, ErrNoJSONTag)
+		}
+
+		// check that the field has the expected Value method
+		valueMethod, ok := fieldIsValue(fieldTy)
+		if !ok {
+			return nil, nil, fmt.Errorf("%s: %w", fieldName, ErrNoValueMethod)
+		}
+
+		ret := valueMethod.Func.Call([]reflect.Value{reflVal.Field(i).Addr()})
+		retValue := ret[0]
+		if ret[1].Interface() != nil {
+			// get the error returned by Value
+			err := ret[1].Interface().(error)
+
+			// we want errors for fields which failed to collect
+			// ErrNotCollectable means that the field is not implemented on the current platform
+			// ignore these errors
+			if !errors.Is(err, ErrNotCollectable) {
+				warns = append(warns, err.Error())
+			}
+
+			// if the field is an error and we don't want to print the default value, continue
+			if !useDefault {
+				continue
+			}
+
+			// use the default value of the type
+			retValue = reflect.Zero(retValue.Type())
+		}
+
+		// try to render the value
+		renderedValue, ok := reflectValueToString(retValue)
+		if !ok {
+			return nil, nil, fmt.Errorf("%s: %w", fieldName, ErrCannotRender)
+		}
+
+		// Get returns an empty string if the key does not exists so no need for particular error handling
+		unit := fieldTy.Tag.Get("unit")
+
+		values[jsonName] = fmt.Sprintf("%s%s", renderedValue, unit)
+	}
+
+	// return an error if no field was successfully collected
+	if len(values) == 0 {
```

Comment:
Do you have a use case in mind for this check as for an empty struct, it makes sense to return an empty map.

Reply:
This is just to keep the behavior we had with the previous API and in the agent.
If no information is collected then it is handled differently in the agent (eg. in the metadata the field is nil and not an empty map, not sure if it changes anything), and I'd rather avoid doing changes which might break something somewhere.